Fire selectionchange event on input and textarea elements

rniwa opened the pull request:
#### 0a064d401bd0a68d148be77d03cb3c157188c04b
<pre>
Fire selectionchange event on input and textarea elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=271033">https://bugs.webkit.org/show_bug.cgi?id=271033</a>

Reviewed by Sihui Liu and Wenson Hsieh.

This PR updates the implementation of selectionchange event to be fired on input element and textarea
element instead of document whenever either real selection or cached selection offsets have been updated
to match the specification: <a href="https://w3c.github.io/selection-api/#selectionchange-event">https://w3c.github.io/selection-api/#selectionchange-event</a>

* LayoutTests/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-bubble-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-expected.txt: Added.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance): Fire selectionchange event on input
element and textarea element when the selection resides within those elements.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::setValueCommon): Fixed a bug that this code was always updating selection
even when TextControlSetValueSelection::DoNotSet is passed in.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::selectionChanged): Made this function return a boolean indicating
whether the selection end points have changed or not.

(WebCore::HTMLTextFormControlElement::scheduleSelectEvent):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Canonical link: <a href="https://commits.webkit.org/276238@main">https://commits.webkit.org/276238@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb6dce60706f2a12359e0882525301191c96e9cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44100 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46533 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46742 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40127 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46405 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20562 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44678 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20196 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17369 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39084 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2150 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48336 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19078 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15643 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43199 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20440 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41926 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20664 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6051 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20067 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -407,8 +407,21 @@ bool FrameSelection::setSelectionWithoutUpdatingAppearance(const VisibleSelectio
+            });
+        }
+    } else {
+        document->eventLoop().queueTask(TaskSource::UserInteraction, [weakDocument = WeakPtr { document.get() }]() {
```
Do we want to fire selectionchangeEvent event when willMutateSelection is false?

rniwa replied:
Oh I guess not. Will fix.

whsieh commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -430,9 +433,21 @@ bool FrameSelection::setSelectionWithoutUpdatingAppearance(const VisibleSelectio
+
+    if (shouldScheduleSelectionChangeEvent) {
+        if (textControl) {
+            document->eventLoop().queueTask(TaskSource::UserInteraction, [ textControl = GCReachableRef { *textControl }]() {
```
```suggestion
            document->eventLoop().queueTask(TaskSource::UserInteraction, [textControl = GCReachableRef { *textControl }] {
```

whsieh replied:
(This extra whitespace still remains)

rniwa replied:
Fixed.

szewai commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -430,9 +433,21 @@ bool FrameSelection::setSelectionWithoutUpdatingAppearance(const VisibleSelectio
+                RefPtr document = weakDocument.get();
+                if (!document)
+                    return;
+                document->dispatchEvent(Event::create(eventNames().selectionchangeEvent, Event::CanBubble::No, Event::IsCancelable::No));
```
Maybe shrink it to:
```
if (RefPtr document = weakDocument.get())
    document->dispatchEvent(Event::create(eventNames().selectionchangeEvent, Event::CanBubble::No, Event::IsCancelable::No));
```

rniwa replied:
Sure, we can do that.

szewai reviewed (approved):
r=me for non-test part; may need to add a test if the behavior change is not covered by any test to import

rniwa committed:
Fire selectionchange event on input and textarea elements
https://bugs.webkit.org/show_bug.cgi?id=271033

Reviewed by Sihui Liu and Wenson Hsieh.

This PR updates the implementation of selectionchange event to be fired on input element and textarea
element instead of document whenever either real selection or cached selection offsets have been updated
to match the specification: https://w3c.github.io/selection-api/#selectionchange-event

* LayoutTests/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-bubble-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/selection/textcontrols/selectionchange-expected.txt: Added.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance): Fire selectionchange event on input
element and textarea element when the selection resides within those elements.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::setValueCommon): Fixed a bug that this code was always updating selection
even when TextControlSetValueSelection::DoNotSet is passed in.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::selectionChanged): Made this function return a boolean indicating
whether the selection end points have changed or not.

(WebCore::HTMLTextFormControlElement::scheduleSelectEvent):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Canonical link: https://commits.webkit.org/276238@main
